src/Microsoft.ML/Runtime/Internal/Tools/CSharpApiGenerator.cs

```diff
@@ -495,16 +495,58 @@ private void GenerateFooter(IndentingTextWriter writer)
+        /// It generates the name based on the <see cref="Type.FullName"/> property of the type
+        /// (see description here https://msdn.microsoft.com/en-us/library/system.type.fullname(v=vs.110).aspx).
+        /// </summary>
+        /// <param name="typesSymbolTable">A dictionary containing the names of the classes already generated.
+        /// This parameter is only used to ensure that the newly generated name is unique.</param>
+        /// <param name="type">The type for which to generate the new name.</param>
+        /// <param name="currentNamespace">The namespace prefix to the new name.</param>
+        /// <returns>A unique name derived from the given type and namespace.</returns>
+        private static string GetSymbolFromType(Dictionary<string, string> typesSymbolTable, Type type, string currentNamespace)
         {
-            var names = typesSymbolTable.Select(kvp => kvp.Value);
-            char dim = fullTypeName.Contains('+') ? '+' : '.';
+            var fullTypeName = type.FullName;
             string name = currentNamespace != "" ? currentNamespace + '.' : "";
 
-            if (fullTypeName.Contains('+'))
-                name += fullTypeName.Substring(0, fullTypeName.LastIndexOf('+')).Substring(fullTypeName.LastIndexOf('.') + 1);
+            int bracketIndex = fullTypeName.IndexOf('[');
+            Type[] genericTypes = null;
+            if (type.IsGenericType)
+                genericTypes = type.GetGenericArguments();
+            if (bracketIndex > 0)
+            {
+                Contracts.AssertValue(genericTypes);
+                fullTypeName = fullTypeName.Substring(0, bracketIndex);
+            }
+
+            // When the type is nested, the names of the outer types are concatenated with a '+'.
+            var nestedNames = fullTypeName.Split('+');
+            var baseName = nestedNames[0];
+            // We currently only handle generic types in the outer most class, support for generic inner classes
+            // can be added if needed.
+            int backTickIndex = baseName.LastIndexOf('`');
+            int dotIndex = baseName.LastIndexOf('.');
+            if (backTickIndex < 0)
+                name += baseName.Substring(dotIndex + 1);
+            else
+            {
+                name += baseName.Substring(dotIndex + 1, backTickIndex - dotIndex - 1);
+                Contracts.AssertValue(genericTypes);
+                if (genericTypes != null)
+                {
+                    foreach (var genType in genericTypes)
+                    {
+                        var splitNames = genType.FullName.Split('+');
+                        if (splitNames[0].LastIndexOf('.') >= 0)
+                            splitNames[0] = splitNames[0].Substring(splitNames[0].LastIndexOf('.') + 1);
+                        name += string.Join("", splitNames);
+                    }
+                }
+            }
```

Comment:
This is great, thanks for doing this. May be put an example of how a name is generated from this control flow.